### PR TITLE
ENH: Improve performance in *stack operations for single array

### PR DESCRIPTION
**Pull request by Kh4L**
#### Current Behavior
- Type checking in `*stack` operations (e.g., `numpy.stack`, `numpy.vstack`, `numpy.hstack`) is currently performed by checking if the input has the `__getitem__` attribute.
- This approach leads to unexpected behavior when the input is an array:
  - Example: `array([2,5,7])` is split into multiple single-element arrays: `(array([2]), array([5]), array([7]))`.

#### Proposed Change
- Treat single array as a sequence of size 1 (a single array), to be compliant with the NumPy documentation: [numpy.stack](https://numpy.org/doc/stable/reference/generated/numpy.stack.html).

This prevents unintended splitting of array inputs, which can lead to confusion and poor performance in user code.

### CODE REVIEW

**Comment by akaashhazarika** [comment]
Looks like some tests are failing with the proposed change.

**Review by jorenham — COMMENTED** [review]
For something to be a "sequence", it often only needs to have a `__getitem__` method that accepts consecutive integers, and sometimes it also needs to have a `__len__` (`_typeshed.SupportsLenAndGetItem`). So there's no need for all of those extra methods that `collections.abc.Sequence` has. 

Contrary to what the typeshed builtin stubs and `collections.abc` suggest, any object with a `__getitem__` method (i.e. the most minimal kind of sequence) can be used in a `for:` loop, and even in all of the builtin functions/constructors that accept "other iterables":

```pycon
>>> class Seq:
...     def __getitem__(self, i: int) -> int:
...         if i > 3:
...             raise IndexError
...         return i
...
>>> for i in Seq():
...     print(i)
... 
0
1
2
3
>>> sum(Seq())
6
```

The runtime behavior of `numpy.stack` is no different:

```pycon
>>> import numpy as np
>>> np.stack(Seq())
array([0, 1, 2, 3])
```

An obvious example of a sequence that isn't a `collection.abc.Sequence` is `numpy.ndarray`. Another one would be `ctypes.Array`, which also doesn't have a `__iter__` method.

TLDR: A `collections.abc.Sequence` is a sequence, but a sequence isn't a `collection.abc.Sequence`.

The failing tests also shows that this additional runtime input restriction will cause backwards-incompatible input restrictions.

**Comment by Kh4L** [comment]
@jorenham thank you for the detailed review.

>This approach leads to unexpected behavior when the input is an array:
>Example: array([2,5,7]) is split into multiple single-element arrays: (array([2]), array([5]), array([7])).

I still believe this can be considered an issue due to the slow performance it yields. I will update the PR to address this while maintaining the use of duck typing.

Please let me know what you think of this new solution

**Comment by jorenham** [comment]
> * This approach leads to unexpected behavior when the input is an array:
>   
>   * Example: `array([2,5,7])` is split into multiple single-element arrays: `(array([2]), array([5]), array([7]))`.

I'm not sure what you mean by this, this is what I'm seeing:

```pycon
>>> import numpy as np
>>> np.__version__
'2.1.1'
>>> a = np.arange(3)
>>> np.stack(a)
array([0, 1, 2])
>>> np.hstack(a)
array([0, 1, 2])
>>> np.vstack(a)
array([[0],
       [1],
       [2]])
```

**Comment by Kh4L** [comment]
Forgot to all the full changes, just pushed.

@jorenham here is a small repro:

```
import numpy as np
import timeit

large_array = np.random.rand(2**18)

def stack_direct():
    np.stack(large_array)

def stack_tuple():
    np.stack((large_array,))

time_direct = timeit.timeit(stack_direct, number=3)
time_tuple = timeit.timeit(stack_tuple, number=3)

print(f"Time for np.stack with array directly: {time_direct:.6f} seconds")
print(f"Time for np.stack with array as a tuple: {time_tuple:.6f} seconds")
```

output:
```
$ python repro.py
Time for np.stack with array directly: 0.454974 seconds
Time for np.stack with array as a tuple: 0.000284 seconds
```

In `stack_direct`, the array is splitted into `2**18` single-element arrays in https://github.com/numpy/numpy/blob/9a2c063342284177e9f51a29f3e616309a3400b1/numpy/_core/shape_base.py#L217

output after the proposed change:

```
$ python repro.py
Time for np.stack with array directly: 0.000282 seconds
Time for np.stack with array as a tuple: 0.000284 seconds
```

**Comment by Kh4L** [comment]
This test is failing after the changes:

However, According to the docs ( https://numpy.org/doc/stable/reference/generated/numpy.stack.html#numpy.stack:~:text=arrayssequence%20of%20array_like ) this should either fail (`TypeError`) or treat the a single `np.array` input as a sequence of array ` . 

I propose changing the test to reflect the behavior proposed above.

@jorenham for feedback

**Comment by jorenham** [comment]
As much as I like this speedup, I'm not sure if the backwards-incompatibility of this change is really worth it 🤔.

But I have to admit that I'm a bit out of my comfortzone here, since this PR has pivoted from something (static) typing-related, to performance/API related...
So it might be wise if we could get a second opinion of another maintainer on this proposed change. Perhaps @charris could help us out here <sub>before we destroy the world</sub>?

**Comment by mattip** [comment]
Documentation building and [a doctest](https://github.com/numpy/numpy/actions/runs/10708274846/job/29690479729?pr=27317#step:8:2918) are failing. Did this somehow change the expected output for a matplotlib use of `stack`? I copy the doc building error since I am not sure you can access the circleci logs.
```
writing output... 
building [html]: targets for 313 source files that are out of date
updating environment: [new config] 2742 added, 0 changed, 0 removed
reading sources... [100%] user/whatisnumpy
/home/circleci/repo/doc/source/user/absolute_beginners.rst:1723: WARNING: Exception occurred in plotting matplotlib3
 from /home/circleci/repo/doc/source/user/plots/matplotlib3.py:
Traceback (most recent call last):
  File "/home/circleci/repo/venv/lib/python3.11/site-packages/matplotlib/sphinxext/plot_directive.py", line 552, in _run_code
    exec(code, ns)
  File "<string>", line 12, in <module>
  File "/home/circleci/repo/venv/lib/python3.11/site-packages/mpl_toolkits/mplot3d/axes3d.py", line 2252, in plot_surface
    polyc = art3d.Poly3DCollection(polys, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/circleci/repo/venv/lib/python3.11/site-packages/mpl_toolkits/mplot3d/art3d.py", line 927, in __init__
    super().__init__(verts, *args, **kwargs)
  File "/home/circleci/repo/venv/lib/python3.11/site-packages/matplotlib/collections.py", line 1200, in __init__
    self.set_verts(verts, closed)
  File "/home/circleci/repo/venv/lib/python3.11/site-packages/mpl_toolkits/mplot3d/art3d.py", line 987, in set_verts
    self._get_vector(verts)
  File "/home/circleci/repo/venv/lib/python3.11/site-packages/mpl_toolkits/mplot3d/art3d.py", line 968, in _get_vector
    self._vec = np.array([xs, ys, zs, ones])
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 2 dimensions. The detected shape was (4, 4) + inhomogeneous part.
looking for now-ou
```

I am not sure why the title refers to type checking, but the actual enhancement is performance related. Also since this is a performance enhancement, please run the [benchmark suite](https://numpy.org/doc/stable/benchmarking.html#numpy-benchmarks), there should be some improvement from the change. If not, please add a new benchmark.

**Comment by mhvk** [comment]
I understand the idea, but worry that it is too much of a breaking change: for multidimensional arrays, `hstack` and `vstack` do not just give back the input via a complicated calculation:
```
In [5]: a = np.arange(24).reshape(2, 3, 4)

In [6]: np.hstack(a)
Out[6]: 
array([[ 0,  1,  2,  3, 12, 13, 14, 15],
       [ 4,  5,  6,  7, 16, 17, 18, 19],
       [ 8,  9, 10, 11, 20, 21, 22, 23]])

In [7]: np.hstack(a).shape
Out[7]: (3, 8)

In [8]: np.vstack(a)
Out[8]: 
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15],
       [16, 17, 18, 19],
       [20, 21, 22, 23]])

In [9]: np.vstack(a).shape
Out[9]: (6, 4)
```

It may be best to just accept this behaviour and make sure it is documented. 

p.s.  It is also not clear returning the input for a single array is necessary best. E.g., if the array is of object dtype and contains multiple arrays inside, I think the present behaviour is preferable to just returning the array.

**Comment by Kh4L** [comment]
@mhvk updated the doc here https://github.com/numpy/numpy/pull/27364

Should we close this PR in favor of #27364 ?

**Comment by mhvk** [comment]
Thanks! Let's indeed close this PR, as I think it is too drastic an API change...
